lib/src/publisher.cc

```diff
@@ -0,0 +1,174 @@
+        std::string assetStr{fty_proto_name(metric)};
+        std::string metricStr{fty_proto_type(metric)};
+
+        //Build the message to send
+        Message msg = Message::buildMessage(
+            "fty-shm",
+            "/etn/metrics/" + assetStr + "/" + metricStr,
+            "MESSAGE",
+            json);
+
+        //Send the message
+        fty::Expected<void> sendRet = getInstance().msgBus->send(msg);
+        if(!sendRet) {
+            logError("Error while sending {}", sendRet.error());
+            return -2;
+        }
+
+        return 0;
+    }
+
+    int Publisher::publishMetric(const std::string& metric, const std::string& asset, const std::string& value, const std::string& unit, uint32_t ttl)
+    {
+        fty_proto_t* proto = protoMetric(metric, asset, value, unit, ttl);
+        int r = publishMetric(proto);
+        fty_proto_destroy(&proto);
+        return r;
+    }
+
+    int Publisher::publishMetric(const std::string& fileName, const std::string& value, const std::string& unit, uint32_t ttl)
+    {
+        // extract metric/asset from fileName (path/to/file/metric@asset)
+        std::string metric{fileName};
+        std::string asset;
+
+        auto pos = metric.rfind("/");
+        if (pos != std::string::npos) {
+            metric = metric.substr(pos + 1);
+        }
+        pos = metric.find("@");
+        if (pos != std::string::npos) {
+            asset = metric.substr(pos + 1);
+            metric = metric.substr(0, pos);
+        }
+
+        return publishMetric(metric, asset, value, unit, ttl);
+    }
+
+    Publisher& Publisher::getInstance()
+    {
+        static Publisher instance; // Guaranteed to be destroyed.
+        return instance;
+    }
+}
+
+// proto metric json serializer
+// returns 0 if success, else <0
+static int metric2JSON(fty_proto_t* metric, std::string& json)
+{
+    json.clear();
+
+    try {
+        if (!metric)
+            { throw std::runtime_error("metric is NULL"); }
+        if (fty_proto_id(metric) != FTY_PROTO_METRIC)
+            { throw std::runtime_error("metric is not FTY_PROTO_METRIC"); }
+
+        const char* type_ = fty_proto_type(metric); // metric
+        const char* name_ = fty_proto_name(metric); // asset name
+        const char* value_ = fty_proto_value(metric);
+        const char* unit_ = fty_proto_unit(metric);
+        const uint32_t ttl_ = fty_proto_ttl(metric);
+
+        std::string metricName{std::string(type_ ? type_ : "null") + "@" + std::string(name_ ? name_ : "null")};
+        std::string value{value_ ? value_ : ""};
+        std::string unit{unit_ ? unit_ : ""};
+        if (unit == " ") unit = ""; // emptied if single space
+
+        cxxtools::SerializationInfo si;
+        si.addMember("metric") <<= metricName;
+        si.addMember("value") <<= value;
+        si.addMember("unit") <<= unit;
+        si.addMember("ttl") <<= ttl_;
+        si.addMember("timestamp") <<= std::to_string(std::time(nullptr)); // epoch time (now)
```

Comment:
ttl is a number, timestamp a string... shall be standardized... as a 'metric' json model/payload